### DISCUSSION TITLE
Scala/Metals and LSP Snippets

### DISCUSSION THREAD

**Discussion by channingwalton** [LazyVim/LazyVim#3498] [Q&A]
Hi,

after all the recent updates I cannot get LSP snippets working with Scala/Metals.

I've tried adding the [LuaSnip Extra](https://www.lazyvim.org/extras/coding/luasnip) but I'm completely lost configuring the various plugins that comes with it.

There is an example metals configuration [here](https://github.com/scalameta/nvim-metals/discussions/39) which uses `vsnip`:

```
snippet = {
          expand = function(args)
            -- Comes from vsnip
            fn["vsnip#anonymous"](args.body)
          end,
        },
```

I guess this is what needs to change to use LuaSnip?

Any help is appreciated.

**Comment by gpwns3717** [comment]
Is there a specfic reason you want to use `vsnip`?
The configuration in metals is an example, and snippets can be used with the current LazyVim config as well.

![SCR-20240607-rzpy](https://github.com/LazyVim/LazyVim/assets/79628639/9da7096a-cabd-4652-95c5-f2afed68071b)

**Reply by channingwalton** [reply]
What I'm trying to get work is LSP snippets, provided by the LSP server, which I _think_ that code block above with vsnip does in the metals example. I just don't know how to do that with the LuaSnip extra.

**Reply by gpwns3717** [reply]
Are there any LSP snippets provided by metals?
I'm not sure, as I've looked through the documentation and can't find them, and I don't think there's been any change, as LazyVim before the release doesn't actually include `vsnip`.

The parts of the old code that included `vsnip` were removed because they weren't valid (`requires` doesn't work in lazy.nvim, and it doesn't add snippets to cmp).

If you think it's still a problem, there is a way to add `vsnip` to your spec.

```lua
return {
  { "garymjr/nvim-snippets", enabled = false },
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      { "hrsh7th/cmp-vsnip" },
      {
        "hrsh7th/vim-vsnip",
        dependencies = { "rafamadriz/friendly-snippets" },
        keys = {
          {
            "<tab>",
            function()
              return vim.fn["vsnip#available"](1) and "<Plug>(vsnip-expand-or-jump)" or "<tab>"
            end,
            expr = true,
            silent = true,
            mode = "i",
          },
          { "<tab>", "<Plug>(vsnip-expand-or-jump)", mode = "s" },
          { "<s-tab>", "<Plug>(vsnip-jump-prev)", mode = { "i", "s" } },
        },
      },
      { "hrsh7th/vim-vsnip-integ" },
    },
    opts = function(_, opts)
      opts.snippet = {
        expand = function(args)
          vim.fn["vsnip#anonymous"](args.body)
        end,
      }
      opts.sources = vim.tbl_filter(function(source)
        return source.name ~= "snippets"
      end, opts.sources)
      table.insert(opts.sources, { name = "vsnip" })
    end,
  },
}
```

**Reply by channingwalton** [reply]
I believe there are snippets, judging by [this](https://github.com/scalameta/metals/blob/1a8f4659f1b17809c02ac77dbe9dd2b479096c31/metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala#L1141).

Thanks for the vsnip suggestion, I'll keep trying.
